alibi_detect/cd/pytorch/lsdd.py

```diff
@@ -0,0 +1,167 @@
+            lambda_rd_max: float = 0.2,
+            device: Optional[str] = None,
+            input_shape: Optional[tuple] = None,
+            data_type: Optional[str] = None
+    ) -> None:
+        """
+        Least-squares density difference (LSDD) data drift detector using a permutation test.
+
+        Parameters
+        ----------
+        x_ref
+            Data used as reference distribution.
+        p_val
+            p-value used for the significance of the permutation test.
+        preprocess_x_ref
+            Whether to already preprocess and store the reference data.
+        update_x_ref
+            Reference data can optionally be updated to the last n instances seen by the detector
+            or via reservoir sampling with size n. For the former, the parameter equals {'last': n} while
+            for reservoir sampling {'reservoir_sampling': n} is passed.
+        preprocess_fn
+            Function to preprocess the data before computing the data drift metrics.
+        sigma
+            Optionally set the GaussianRBF kernel bandwidth. Can also pass multiple bandwidth values as an array.
+            The kernel evaluation is then averaged over those bandwidths.
+        n_permutations
+            Number of permutations used in the permutation test.
+        n_kernel_centers
+            Number of reference data points to use kernel centers to use in the estimation of the LSDD.
+            Defaults to 1/20th of the reference data.
+        lambda_rd_max
+            The maximum relative difference between two estimates of LSDD that the regularization parameter
+            lambda is allowed to cause. Defaults to 0.2 as in the paper.
+        device
+            Device type used. The default None tries to use the GPU and falls back on CPU if needed.
+            Can be specified by passing either 'cuda', 'gpu' or 'cpu'.
+        input_shape
+            Shape of input data.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+        super().__init__(
+            x_ref=x_ref,
+            p_val=p_val,
+            preprocess_x_ref=preprocess_x_ref,
+            update_x_ref=update_x_ref,
+            preprocess_fn=preprocess_fn,
+            sigma=sigma,
+            n_permutations=n_permutations,
+            n_kernel_centers=n_kernel_centers,
+            lambda_rd_max=lambda_rd_max,
+            input_shape=input_shape,
+            data_type=data_type
+        )
+        self.meta.update({'backend': 'pytorch'})
+
+        # set backend
+        if device is None or device.lower() in ['gpu', 'cuda']:
+            self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+            if self.device.type == 'cpu':
+                print('No GPU detected, fall back on CPU.')
+        else:
+            self.device = torch.device('cpu')
+
+        if self.preprocess_x_ref or self.preprocess_fn is None:
```

Comment:
There is no device management during init, is this fine?

Reply:
Sorry, I don't follow what you mean here?

Reply:
During init, e.g. `x_ref = torch.as_tensor(self.x_ref)`, `_ = self.kernel(x_ref, x_ref, infer_sigma=True)` or `
            self.H = GaussianRBF(np.sqrt(2.)*self.kernel.sigma)(self.kernel_centers, self.kernel_centers) * \                ((np.pi*self.kernel.sigma**2)**(d/2))` PyTorch tensors are being used and the kernel is initialized but it is all assumed to be on CPU, regardless whether the GPU is specified when initializing the detector.

Reply:
Sorted for both offline and online lsdd